test/storage.test.js

```diff
@@ -129,6 +132,107 @@ describe('storage.js', () => {
+            };
+            
+            mockFs.restore();
+            mockRequire('../lib/storage/datauri', {
+                getPreSignedUrl : (source) => {
+                    console.log('Mocked source : ' + source);
+                    return 'https://example.com/preSignedUrl';
+                },
+                download : () => {
+                    console.log('Fake file downloaded');
+                }
+            });
+            
+            const directory = './in/fakeSource/filePath';
+            const name = 'source';
+            const disableSourceDownload = true;
+
+            const {getAsset} = mockRequire.reRequire("../lib/storage");
+            const source = await getAsset(assetReference, directory, name, disableSourceDownload);
+            assert.strictEqual(source.name, 'source');
+            assert.strictEqual(source.path, 'in/fakeSource/filePath/source');
+            assert.strictEqual(source.params.url, 'http://nondatauriurl.com/example.png');
+        });
+
+        it("should fail to generate preSignUrl for  empty data uri", async() =>{
+            const assetReference = {
+                url: "data:,"
+            };
+            const directory = './in/fakeSource/filePath';
+            const name = 'source';
+            const disableSourceDownload = true;
+            mockFs({ './in/fakeSource/filePath': {} });
+            assert.ok(fs.existsSync(directory));
+            try {
+                await getAsset(assetReference, directory, name, disableSourceDownload);
+            } catch (e) {
+                assert.strictEqual(e.name, 'SourceUnsupportedError');
+                assert.strictEqual(e.message, 'Invalid or missing local file in/fakeSource/filePath/source');
+            }
+        });
+
+        it("should fail to generate preSignUrl for non existent file", async() =>{
+            const assetReference = {
+                url: "data:,"
+            };
+            const directory = './in/fakeSource/filePath';
+            const name = 'source';
+            const disableSourceDownload = true;
+            mockFs.restore();
+            mockRequire('../lib/storage/datauri', {
+                getPreSignedUrl : (source) => {
+                    console.log('Mocked source : ' + source);
+                    return 'https://example.com/preSignedUrl';
+                },
+                download : () => {
+                    console.log('Fake file downloaded');
+                }
+            });
+            mockFs({ './in/fakeSource/filePath': {} });
+            assert.ok(fs.existsSync(directory));
+            try {
+                await getAsset(assetReference, directory, name, disableSourceDownload);
+            } catch (e) {
+                assert.strictEqual(e.name, 'SourceUnsupportedError');
+                assert.strictEqual(e.message, 'Invalid or missing local file in/fakeSource/filePath/source');
+            }
+        });
```

Comment:
maybe add a test case to validate the scenario when aio-lib-files throws an exception

Reply:
(you can use mock-require to define a behavior that will throw)

Reply:
Added GenericError and mocks